File: backend/src/db/repository.ts
import { supabase } from "../config/supabase.js";
import type { IVideo, VideoSource, VideoStatus } from "../models/Video.js";
import type { ITranscript, ISegment } from "../models/Transcript.js";
import type { IGraph, IGraphEdge, IGraphNode } from "../models/Graph.js";
import type { ContentType, IGeneratedContent } from "../models/GeneratedContent.js";

function fail(error: { message: string } | null): void {
  if (error) throw new Error(error.message);
}

function mapVideo(row: any): IVideo {
  return {
    _id: row.id,
    source: row.source,
    originalName: row.original_name,
    url: row.url ?? undefined,
    filePath: row.file_path ?? undefined,
    duration: Number(row.duration ?? 0),
    status: row.status,
    owner: row.owner_id,
    errorMessage: row.error_message ?? undefined,
    targetLanguage: row.target_language ?? "en",
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export async function createVideo(input: {
  source: VideoSource;
  originalName: string;
  ownerId: string;
  url?: string;
  filePath?: string;
  targetLanguage?: string;
}): Promise<IVideo> {
  const { data, error } = await supabase
    .from("videos")
    .insert({
      source: input.source,
      original_name: input.originalName,
      owner_id: input.ownerId,
      url: input.url ?? null,
      file_path: input.filePath ?? null,
      target_language: input.targetLanguage ?? "en",
      status: "queued",
    })
    .select("*")
    .single();
  fail(error);
  return mapVideo(data);
}

export async function findVideo(id: string, ownerId?: string): Promise<IVideo | null> {
  let query = supabase.from("videos").select("*").eq("id", id);
  if (ownerId) query = query.eq("owner_id", ownerId);
  const { data, error } = await query.maybeSingle();
  fail(error);
  return data ? mapVideo(data) : null;
}

export async function listVideos(ownerId: string): Promise<IVideo[]> {
  const { data, error } = await supabase
    .from("videos")
    .select("*")
    .eq("owner_id", ownerId)
    .order("created_at", { ascending: false });
  fail(error);
  return (data ?? []).map(mapVideo);
}

// Used by the /api/files guard: the requested file must belong to one of the
// caller's videos (matched by stored path suffix), otherwise it is 404'd.
export async function findVideoByFileName(fileName: string, ownerId: string): Promise<IVideo | null> {
  const { data, error } = await supabase
    .from("videos")
    .select("*")
    .eq("owner_id", ownerId)
    .like("file_path", `%/${fileName}`)
    .maybeSingle();
  fail(error);
  return data ? mapVideo(data) : null;
}

export async function updateVideo(
  id: string,
  patch: Partial<Pick<IVideo, "status" | "filePath" | "duration" | "originalName">> & { errorMessage?: string | null }
): Promise<IVideo> {
  const mapped: Record<string, unknown> = {};
  if (patch.status !== undefined) mapped.status = patch.status;
  if (patch.filePath !== undefined) mapped.file_path = patch.filePath;
  if (patch.duration !== undefined) mapped.duration = patch.duration;
  if (patch.originalName !== undefined) mapped.original_name = patch.originalName;
  if (patch.errorMessage !== undefined) mapped.error_message = patch.errorMessage;
  const { data, error } = await supabase.from("videos").update(mapped).eq("id", id).select("*").single();
  fail(error);
  return mapVideo(data);
}

function mapSegment(row: any): ISegment {
  return {
    start: Number(row.start_time),
    end: Number(row.end_time),
    speaker: row.speaker ?? "",
    language: row.language,
    text: row.text,
    confidence: Number(row.confidence ?? 0),
  };
}

export async function getTranscript(videoId: string): Promise<ITranscript | null> {
  const { data: transcript, error } = await supabase
    .from("transcripts")
    .select("*")
    .eq("video_id", videoId)
    .maybeSingle();
  fail(error);
  if (!transcript) return null;
  const { data: segments, error: segmentError } = await supabase
    .from("transcript_segments")
    .select("*")
    .eq("transcript_id", transcript.id)
    .order("position", { ascending: true });
  fail(segmentError);
  return {
    _id: transcript.id,
    videoId,
    language: transcript.language,
    segments: (segments ?? []).map(mapSegment),
    createdAt: transcript.created_at,
    updatedAt: transcript.updated_at,
  };
}

export async function upsertTranscript(videoId: string, language: string, segments: ISegment[]): Promise<void> {
  const { data: transcript, error } = await supabase
    .from("transcripts")
    .upsert({ video_id: videoId, language }, { onConflict: "video_id" })
    .select("id")
    .single();
  fail(error);
  if (!transcript) throw new Error("Transcript upsert returned no row");
  const { error: deleteError } = await supabase.from("transcript_segments").delete().eq("transcript_id", transcript.id);
  fail(deleteError);
  const rows = segments.map((segment, position) => ({
    transcript_id: transcript.id,
    position,
    start_time: segment.start,
    end_time: segment.end,
    speaker: segment.speaker ?? "",
    language: segment.language,
    text: segment.text,
    confidence: segment.confidence ?? 0,
  }));
  for (let i = 0; i < rows.length; i += 500) {
    const { error: insertError } = await supabase.from("transcript_segments").insert(rows.slice(i, i + 500));
    fail(insertError);
  }
}

function mapGraph(videoId: string, row: any, nodes: any[], edges: any[]): IGraph {
  const mappedNodes: IGraphNode[] = nodes.map((node) => ({
    id: node.node_id,
    label: node.label,
    type: node.type,
    timestampRef: node.timestamp_ref == null ? null : Number(node.timestamp_ref),
    summary: node.summary ?? null,
    metadata: node.metadata ?? {},
  }));
  const mappedEdges: IGraphEdge[] = edges.map((edge) => ({
    source: edge.source_id,
    target: edge.target_id,
    relation: edge.relation,
  }));
  return { _id: row.id, videoId, nodes: mappedNodes, edges: mappedEdges, createdAt: row.created_at, updatedAt: row.updated_at };
}

export async function getGraph(videoId: string): Promise<IGraph | null> {
  const { data: graph, error } = await supabase.from("graphs").select("*").eq("video_id", videoId).maybeSingle();
  fail(error);
  if (!graph) return null;
  const [{ data: nodes, error: nodeError }, { data: edges, error: edgeError }] = await Promise.all([
    supabase.from("graph_nodes").select("*").eq("graph_id", graph.id).order("created_at"),
    supabase.from("graph_edges").select("*").eq("graph_id", graph.id).order("created_at"),
  ]);
  fail(nodeError);
  fail(edgeError);
  return mapGraph(videoId, graph, nodes ?? [], edges ?? []);
}

export async function upsertGraph(videoId: string, nodes: IGraphNode[], edges: IGraphEdge[]): Promise<void> {
  const { data: graph, error } = await supabase.from("graphs").upsert({ video_id: videoId }, { onConflict: "video_id" }).select("id").single();
  fail(error);
  if (!graph) throw new Error("Graph upsert returned no row");
  const [{ error: nodeDeleteError }, { error: edgeDeleteError }] = await Promise.all([
    supabase.from("graph_nodes").delete().eq("graph_id", graph.id),
    supabase.from("graph_edges").delete().eq("graph_id", graph.id),
  ]);
  fail(nodeDeleteError);
  fail(edgeDeleteError);
  if (nodes.length) {
    const { error: nodeError } = await supabase.from("graph_nodes").insert(nodes.map((node) => ({
      graph_id: graph.id,
      node_id: node.id,
      label: node.label,
      type: node.type,
      timestamp_ref: node.timestampRef,
      summary: node.summary,
      metadata: node.metadata,
    })));
    fail(nodeError);
  }
  if (edges.length) {
    const { error: edgeError } = await supabase.from("graph_edges").insert(edges.map((edge) => ({
      graph_id: graph.id,
      source_id: edge.source,
      target_id: edge.target,
      relation: edge.relation,
    })));
    fail(edgeError);
  }
}

function mapGenerated(row: any): IGeneratedContent {
  return { _id: row.id, videoId: row.video_id, type: row.type, content: row.content, format: row.format, createdAt: row.created_at, updatedAt: row.updated_at };
}

export async function getGenerated(videoId: string, type?: ContentType): Promise<IGeneratedContent[]> {
  let query = supabase.from("generated_content").select("*").eq("video_id", videoId).order("created_at", { ascending: false });
  if (type) query = query.eq("type", type);
  const { data, error } = await query;
  fail(error);
  return (data ?? []).map(mapGenerated);
}

export async function upsertGenerated(videoId: string, type: ContentType, content: string, format = "markdown"): Promise<IGeneratedContent> {
  const { data, error } = await supabase.from("generated_content").upsert({ video_id: videoId, type, content, format }, { onConflict: "video_id,type" }).select("*").single();
  fail(error);
  return mapGenerated(data);
}

File: backend/app/db/repository.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel
from app.db.supabase import supabase
from app.schemas.schemas import SegmentOut, GraphNode, GraphEdge, ChapterOut

def fail(response):
    # supabase-py v2.0+ generally raises exceptions on error if the request fails, 
    # but we can check if data is returned appropriately.
    pass

def create_video(source: str, original_name: str, owner_id: str, url: str = None, file_path: str = None, target_language: str = "en") -> dict:
    data = {
        "source": source,
        "original_name": original_name,
        "owner_id": owner_id,
        "target_language": target_language,
        "status": "queued"
    }
    if url:
        data["url"] = url
    if file_path:
        data["file_path"] = file_path

    response = supabase.table("videos").insert(data).execute()
    return response.data[0]

def find_video(video_id: str, owner_id: Optional[str] = None) -> Optional[dict]:
    query = supabase.table("videos").select("*").eq("id", video_id)
    if owner_id:
        query = query.eq("owner_id", owner_id)
    response = query.execute()
    return response.data[0] if response.data else None

def list_videos(owner_id: str) -> List[dict]:
    response = supabase.table("videos").select("*").eq("owner_id", owner_id).order("created_at", desc=True).execute()
    return response.data

def find_video_by_file_name(file_name: str, owner_id: str) -> Optional[dict]:
    response = supabase.table("videos").select("*").eq("owner_id", owner_id).ilike("file_path", f"%/{file_name}").execute()
    return response.data[0] if response.data else None

def update_video(video_id: str, payload: dict) -> dict:
    res = supabase.table("videos").update(payload).eq("id", video_id).execute()
    return res.data[0] if res.data else None

def delete_video(video_id: str, owner_id: str) -> list:
    res = supabase.table("videos").delete().eq("id", video_id).eq("owner_id", owner_id).execute()
    return res.data

def get_transcript(video_id: str) -> Optional[dict]:
    transcript_res = supabase.table("transcripts").select("*").eq("video_id", video_id).execute()
    if not transcript_res.data:
        return None
    transcript = transcript_res.data[0]
    segments_res = supabase.table("transcript_segments").select("*").eq("transcript_id", transcript["id"]).order("position").execute()
    return {
        "_id": transcript["id"],
        "videoId": video_id,
        "language": transcript["language"],
        "segments": [{
            "start": s["start_time"],
            "end": s["end_time"],
            "speaker": s.get("speaker", ""),
            "language": s["language"],
            "text": s["text"],
            "confidence": s.get("confidence", 0.0)
        } for s in segments_res.data]
    }

def upsert_transcript(video_id: str, language: str, segments: List[SegmentOut]) -> None:
    transcript_res = supabase.table("transcripts").upsert({"video_id": video_id, "language": language}, on_conflict="video_id").execute()
    if not transcript_res.data:
        raise ValueError("Transcript upsert failed")
    transcript_id = transcript_res.data[0]["id"]
    supabase.table("transcript_segments").delete().eq("transcript_id", transcript_id).execute()
    
    rows = []
    for pos, seg in enumerate(segments):
        rows.append({
            "transcript_id": transcript_id,
            "position": pos,
            "start_time": seg.start,
            "end_time": seg.end,
            "speaker": seg.speaker or "",
            "language": seg.language,
            "text": seg.text,
            "confidence": seg.confidence or 0.0
        })
    for i in range(0, len(rows), 500):
        supabase.table("transcript_segments").insert(rows[i:i+500]).execute()

def get_graph(video_id: str) -> Optional[dict]:
    graph_res = supabase.table("graphs").select("*").eq("video_id", video_id).execute()
    if not graph_res.data:
        return None
    graph_id = graph_res.data[0]["id"]
    nodes_res = supabase.table("graph_nodes").select("*").eq("graph_id", graph_id).order("created_at").execute()
    edges_res = supabase.table("graph_edges").select("*").eq("graph_id", graph_id).order("created_at").execute()
    
    return {
        "_id": graph_id,
        "videoId": video_id,
        "nodes": [{
            "id": n["node_id"],
            "label": n["label"],
            "type": n["type"],
            "timestampRef": n.get("timestamp_ref"),
            "summary": n.get("summary"),
            "metadata": n.get("metadata", {})
        } for n in nodes_res.data],
        "edges": [{
            "source": e["source_id"],
            "target": e["target_id"],
            "relation": e["relation"]
        } for e in edges_res.data]
    }

def upsert_graph(video_id: str, nodes: List[GraphNode], edges: List[GraphEdge]) -> None:
    graph_res = supabase.table("graphs").upsert({"video_id": video_id}, on_conflict="video_id").execute()
    graph_id = graph_res.data[0]["id"]
    supabase.table("graph_nodes").delete().eq("graph_id", graph_id).execute()
    supabase.table("graph_edges").delete().eq("graph_id", graph_id).execute()
    
    if nodes:
        node_rows = [{
            "graph_id": graph_id,
            "node_id": n.id,
            "label": n.label,
            "type": n.type,
            "timestamp_ref": n.timestampRef,
            "summary": n.summary,
            "metadata": n.metadata
        } for n in nodes]
        for i in range(0, len(node_rows), 500):
            supabase.table("graph_nodes").insert(node_rows[i:i+500]).execute()
            
    if edges:
        edge_rows = [{
            "graph_id": graph_id,
            "source_id": e.source,
            "target_id": e.target,
            "relation": e.relation
        } for e in edges]
        for i in range(0, len(edge_rows), 500):
            supabase.table("graph_edges").insert(edge_rows[i:i+500]).execute()

def get_generated(video_id: str, type: Optional[str] = None) -> List[dict]:
    query = supabase.table("generated_content").select("*").eq("video_id", video_id).order("created_at", desc=True)
    if type:
        query = query.eq("type", type)
    res = query.execute()
    return [{"_id": r["id"], "videoId": r["video_id"], "type": r["type"], "content": r["content"], "format": r["format"]} for r in res.data]

def upsert_generated(video_id: str, type: str, content: str, format: str = "markdown") -> dict:
    res = supabase.table("generated_content").upsert({
        "video_id": video_id,
        "type": type,
        "content": content,
        "format": format
    }, on_conflict="video_id,type").execute()
    r = res.data[0]
    return {"_id": r["id"], "videoId": r["video_id"], "type": r["type"], "content": r["content"], "format": r["format"]}

def get_chapters(video_id: str) -> List[ChapterOut]:
    res = supabase.table("video_chapters").select("*").eq("video_id", video_id).order("chapter_index").execute()
    return [ChapterOut(
        id=r["id"],
        videoId=r["video_id"],
        title=r["title"],
        start=float(r["start_time"]),
        end=float(r["end_time"]),
        summary=r.get("summary", ""),
        keywords=r.get("keywords", [])
    ) for r in res.data]

def upsert_chapters(video_id: str, chapters: List[ChapterOut]) -> None:
    if not chapters:
        return
    rows = []
    for i, c in enumerate(chapters):
        rows.append({
            "video_id": video_id,
            "chapter_index": i,
            "title": c.title,
            "start_time": c.start,
            "end_time": c.end,
            "summary": c.summary or "",
            "keywords": c.keywords or []
        })
    supabase.table("video_chapters").upsert(rows, on_conflict="video_id,chapter_index").execute()
